pkg/providers/tinkerbell/comparator.go

```diff
@@ -0,0 +1,51 @@
+package tinkerbell
+
+import (
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+)
+
+// VersionComparable enables kubernetes version retrieval.
+type VersionComparable interface {
+	Version() string
+}
+
+// MachineReplicaComparable enables replica number retrieval.
+type MachineReplicaComparable interface {
+	Replicas() int
+}
+
+// KubeadmControlPlaneComparator wraps the KubeadmControlPlane to enable Comparator functions.
+type KubeadmControlPlaneComparator struct {
+	*controlplanev1.KubeadmControlPlane
+}
+
+// Version retrieves KubeadmControlPlane replicas.
+func (k *KubeadmControlPlaneComparator) Version() string {
+	return k.Spec.Version
+}
+
+// Replicas retrieves KubeadmControlPlane replicas.
+func (k *KubeadmControlPlaneComparator) Replicas() int {
+	return int(*k.Spec.Replicas)
+}
+
+// MachineDeploymentComparator implements MachineReplicaComparable by decorating a MachineDeployment.
+type MachineDeploymentComparator struct {
+	*clusterv1.MachineDeployment
+}
+
+// Replicas retrieves MachineDeployment replicas.
+func (m *MachineDeploymentComparator) Replicas() int {
+	return int(*m.Spec.Replicas)
+}
+
+// HasKubernetesVersionChange detects kubernetes version has changed between two Comparator objects.
+func HasKubernetesVersionChange(oldObj VersionComparable, newObj VersionComparable) bool {
```

Comment:
where are these two methods going to be used?

Reply:
This would be to detect the Tinkerbell Upgrade Operation, whether it's rolling or scaling. So these will be to compare the CAPI current vs desired state via KubeadmControlPlane for control plane and MachineDeployment for worker nodes.